File: apple/Packages/DsmNetwork/Sources/DsmQuickConnectResolver.swift
import CryptoKit
import DsmCore
import Foundation
import DsmLocalization

public enum QuickConnectEndpointKind: Equatable, Sendable {
    case local
    case external
    case relay
}

public struct QuickConnectEndpoint: Equatable, Sendable {
    public let host: String
    public let port: Int
    public let kind: QuickConnectEndpointKind

    public init(host: String, port: Int, kind: QuickConnectEndpointKind = .external) {
        self.host = host
        self.port = port
        self.kind = kind
    }
}

public enum QuickConnectResolutionError: Error, Equatable, Sendable {
    case notFound
    case offline
    case noDirectRoute
    case relayDisabled
    case relayUnavailable
    case relayIdentityMismatch
    case serviceUnavailable
    case invalidResponse
}

extension QuickConnectResolutionError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .notFound:
            L10n.string("shared.d96486036c975839")
        case .offline:
            L10n.string("shared.d2273a40a6f9387a")
        case .noDirectRoute:
            L10n.string("shared.2952c5ce0efa0ffd")
        case .relayDisabled:
            L10n.string("shared.ef6c11be2ae49b06")
        case .relayUnavailable:
            L10n.string("shared.9e73a1cad5867f75")
        case .relayIdentityMismatch:
            L10n.string("shared.9edd5ea14dc200aa")
        case .serviceUnavailable:
            L10n.string("shared.25e80b22c8d9f4e6")
        case .invalidResponse:
            L10n.string("shared.b0738d49115e3229")
        }
    }
}

public protocol QuickConnectResolving: Sendable {
    func resolve(id: String) async throws -> [QuickConnectEndpoint]
    func requestRelay(id: String) async throws -> QuickConnectEndpoint
}

private struct QuickConnectCommand: Encodable {
    let version = 1
    let command: String
    let stopWhenError = false
    let stopWhenSuccess: Bool
    let id = "mainapp_https"
    let serverID: String
    let isGofile = false
    let path = ""

    private enum CodingKeys: String, CodingKey {
        case version
        case command
        case stopWhenError = "stop_when_error"
        case stopWhenSuccess = "stop_when_success"
        case id
        case serverID
        case isGofile = "is_gofile"
        case path
    }
}

private struct QuickConnectResponse: Decodable {
    let errno: Int
    let server: Server?
    let service: Service?
    let smartDNS: SmartDNS?

    struct Server: Decodable {
        let state: String?
        let serverID: String?
        let pingpongPath: String?

        private enum CodingKeys: String, CodingKey {
            case state = "ds_state"
            case serverID
            case pingpongPath = "pingpong_path"
        }
    }

    struct Service: Decodable {
        let port: Int?
        let relayIP: String?
        let relayPort: Int?

        private enum CodingKeys: String, CodingKey {
            case port
            case relayIP = "relay_ip"
            case relayPort = "relay_port"
        }
    }

    struct SmartDNS: Decodable {
        let host: String?
        let lan: [String]?
    }

    struct Environment: Decodable {
        let controlHost: String?
        let relayRegion: String?

        private enum CodingKeys: String, CodingKey {
            case controlHost = "control_host"
            case relayRegion = "relay_region"
        }
    }

    let environment: Environment?

    private enum CodingKeys: String, CodingKey {
        case errno
        case server
        case service
        case smartDNS = "smartdns"
        case environment = "env"
    }
}

private struct QuickConnectPingPongResponse: Decodable {
    let ezid: String
}

struct QuickConnectRelayDescriptor {
    let endpoint: QuickConnectEndpoint
    let serverID: String
    let pingpongPath: String
}

public actor DsmQuickConnectResolver: QuickConnectResolving {
    private let session: URLSession
    private let controlURLs: [URL]
    private let maximumResponseBytes: Int

    public init() {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 10
        configuration.timeoutIntervalForResource = 15
        configuration.urlCache = nil
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        session = URLSession(configuration: configuration)

        let isChina = Locale.current.region?.identifier.uppercased() == "CN"
        let domains = isChina
            ? ["global.quickconnect.cn", "global.quickconnect.to"]
            : ["global.quickconnect.to", "global.quickconnect.cn"]
        controlURLs = domains.compactMap { URL(string: "https://\($0)/Serv.php") }
        maximumResponseBytes = 1_024 * 1_024
    }

    init(
        session: URLSession,
        controlURLs: [URL],
        maximumResponseBytes: Int = 1_024 * 1_024
    ) {
        self.session = session
        self.controlURLs = controlURLs
        self.maximumResponseBytes = max(maximumResponseBytes, 1)
    }

    public func resolve(id: String) async throws -> [QuickConnectEndpoint] {
        guard NasAddressParser.isPotentialQuickConnectID(id) else {
            throw QuickConnectResolutionError.notFound
        }

        var lastError: Error = QuickConnectResolutionError.serviceUnavailable

        for controlURL in controlURLs {
            do {
                let data = try await send(
                    command: "get_server_info",
                    stopWhenSuccess: false,
                    serverID: id,
                    to: controlURL
                )
                return try Self.decodeEndpoints(from: data)
            } catch let error as QuickConnectResolutionError {
                lastError = error
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                lastError = QuickConnectResolutionError.serviceUnavailable
            }
        }

        throw lastError
    }

    public func requestRelay(id: String) async throws -> QuickConnectEndpoint {
        guard NasAddressParser.isPotentialQuickConnectID(id) else {
            throw QuickConnectResolutionError.notFound
        }

        let controlURL = try await resolveControlURL(id: id)
        var lastError: Error = QuickConnectResolutionError.relayUnavailable

        for attempt in 0..<3 {
            do {
                // Synology 内部 API：公开白皮书描述了中继流程，但未公开此控制请求契约。
                let data = try await send(
                    command: "request_tunnel",
                    stopWhenSuccess: true,
                    serverID: id,
                    to: controlURL
                )
                let descriptor = try Self.decodeRelayDescriptor(from: data, quickConnectID: id)
                try await verifyRelay(descriptor)
                return descriptor.endpoint
            } catch let error as QuickConnectResolutionError where error == .relayDisabled {
                throw error
            } catch let error as QuickConnectResolutionError where error == .relayIdentityMismatch {
                // 身份不一致属于安全失败，不应通过重新申请中继来自动绕过。
                throw error
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                lastError = error
                if attempt < 2 {
                    try await Task.sleep(for: .seconds(attempt + 1))
                }
            }
        }

        if let error = lastError as? QuickConnectResolutionError,
           error == .relayIdentityMismatch {
            throw error
        }
        throw QuickConnectResolutionError.relayUnavailable
    }

    private func resolveControlURL(id: String) async throws -> URL {
        var lastError: Error = QuickConnectResolutionError.serviceUnavailable
        for controlURL in controlURLs {
            do {
                let data = try await send(
                    command: "get_server_info",
                    stopWhenSuccess: false,
                    serverID: id,
                    to: controlURL
                )
                let host = try Self.decodeControlHost(from: data)
                guard let url = URL(string: "https://\(host)/Serv.php") else {
                    throw QuickConnectResolutionError.invalidResponse
                }
                return url
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                lastError = error
            }
        }
        throw lastError
    }

    private func send(
        command: String,
        stopWhenSuccess: Bool,
        serverID: String,
        to controlURL: URL
    ) async throws -> Data {
        let body = try JSONEncoder().encode([
            QuickConnectCommand(
                command: command,
                stopWhenSuccess: stopWhenSuccess,
                serverID: serverID
            )
        ])
        var request = URLRequest(url: controlURL)
        request.httpMethod = "POST"
        request.timeoutInterval = command == "request_tunnel" ? 30 : 10
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = body

        let (data, response) = try await boundedData(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200..<300).contains(httpResponse.statusCode) else {
            throw QuickConnectResolutionError.serviceUnavailable
        }
        return data
    }

    private func verifyRelay(_ descriptor: QuickConnectRelayDescriptor) async throws {
        guard descriptor.pingpongPath.hasPrefix("/"),
              descriptor.pingpongPath.count <= 2_048,
              !descriptor.pingpongPath.contains("://"),
              !descriptor.pingpongPath.contains("#"),
              let url = URL(
                string: "https://\(descriptor.endpoint.host)\(descriptor.pingpongPath)"
              ),
              url.host?.lowercased() == descriptor.endpoint.host else {
            throw QuickConnectResolutionError.invalidResponse
        }

        let expectedID = Insecure.MD5.hash(data: Data(descriptor.serverID.utf8))
            .map { String(format: "%02x", $0) }
            .joined()

        for attempt in 0..<6 {
            do {
                var request = URLRequest(url: url)
                request.timeoutInterval = 15
                request.setValue("application/json", forHTTPHeaderField: "Accept")
                let (data, response) = try await boundedData(for: request)
                guard let httpResponse = response as? HTTPURLResponse,
                      (200..<300).contains(httpResponse.statusCode) else {
                    throw QuickConnectResolutionError.relayUnavailable
                }
                let pingPong = try JSONDecoder().decode(QuickConnectPingPongResponse.self, from: data)
                guard pingPong.ezid.lowercased() == expectedID else {
                    throw QuickConnectResolutionError.relayIdentityMismatch
                }
                return
            } catch let error as QuickConnectResolutionError where error == .relayIdentityMismatch {
                throw error
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                if attempt < 5 {
                    try await Task.sleep(for: .seconds(1))
                }
            }
        }
        throw QuickConnectResolutionError.relayUnavailable
    }

    /// QuickConnect 控制响应必须在累积前受限，不能先由 data(for:) 无界缓冲。
    private func boundedData(
        for request: URLRequest
    ) async throws -> (Data, URLResponse) {
        let (bytes, response) = try await session.bytes(for: request)
        guard response.expectedContentLength <= Int64(maximumResponseBytes) else {
            throw QuickConnectResolutionError.invalidResponse
        }
        var data = Data()
        if response.expectedContentLength > 0 {
            data.reserveCapacity(
                min(Int(response.expectedContentLength), maximumResponseBytes)
            )
        }
        for try await byte in bytes {
            guard data.count < maximumResponseBytes else {
                throw QuickConnectResolutionError.invalidResponse
            }
            data.append(byte)
        }
        return (data, response)
    }

    static func decodeEndpoints(from data: Data) throws -> [QuickConnectEndpoint] {
        let responses: [QuickConnectResponse]
        do {
            responses = try JSONDecoder().decode([QuickConnectResponse].self, from: data)
        } catch {
            throw QuickConnectResolutionError.invalidResponse
        }

        guard let response = responses.first(where: { $0.errno == 0 }) else {
            throw QuickConnectResolutionError.notFound
        }
        guard response.server?.state?.uppercased() == "CONNECTED" else {
            throw QuickConnectResolutionError.offline
        }
        guard let port = response.service?.port,
              (1...65_535).contains(port) else {
            throw QuickConnectResolutionError.noDirectRoute
        }

        var seen = Set<String>()
        let local = (response.smartDNS?.lan ?? []).compactMap { host -> QuickConnectEndpoint? in
            let normalized = host.lowercased()
            guard Self.isTrustedDirectHost(normalized), seen.insert(normalized).inserted else {
                return nil
            }
            return QuickConnectEndpoint(host: normalized, port: port, kind: .local)
        }
        let external = [response.smartDNS?.host].compactMap { host -> QuickConnectEndpoint? in
            guard let host else {
                return nil
            }
            let normalized = host.lowercased()
            guard Self.isTrustedDirectHost(normalized), seen.insert(normalized).inserted else {
                return nil
            }
            return QuickConnectEndpoint(host: normalized, port: port, kind: .external)
        }
        let endpoints = local + external
        guard !endpoints.isEmpty else {
            throw QuickConnectResolutionError.noDirectRoute
        }
        return endpoints
    }

    static func decodeControlHost(from data: Data) throws -> String {
        let responses = try decodeResponses(from: data)
        guard let response = responses.first(where: { $0.errno == 0 }),
              response.server?.state?.uppercased() == "CONNECTED",
              let host = response.environment?.controlHost?.lowercased(),
              isTrustedControlHost(host) else {
            throw QuickConnectResolutionError.invalidResponse
        }
        return host
    }

    static func decodeRelayDescriptor(
        from data: Data,
        quickConnectID: String
    ) throws -> QuickConnectRelayDescriptor {
        let responses = try decodeResponses(from: data)
        if responses.contains(where: { $0.errno == 19 }) {
            throw QuickConnectResolutionError.relayDisabled
        }
        guard let response = responses.first(where: { $0.errno == 0 }),
              let relayIP = response.service?.relayIP,
              !relayIP.isEmpty,
              let relayPort = response.service?.relayPort,
              (1...65_535).contains(relayPort),
              let serverID = response.server?.serverID,
              !serverID.isEmpty,
              let region = response.environment?.relayRegion?.lowercased(),
              isValidHostLabel(region),
              let controlHost = response.environment?.controlHost?.lowercased(),
              isTrustedControlHost(controlHost) else {
            throw QuickConnectResolutionError.relayUnavailable
        }

        let topDomain = controlHost.hasSuffix(".quickconnect.cn") ? "cn" : "to"
        let host = "\(quickConnectID.lowercased()).\(region).quickconnect.\(topDomain)"
        guard isTrustedRelayHost(host) else {
            throw QuickConnectResolutionError.invalidResponse
        }
        let pingpongPath = response.server?.pingpongPath
            ?? "/webman/pingpong.cgi?action=cors&quickconnect=true"
        return QuickConnectRelayDescriptor(
            endpoint: QuickConnectEndpoint(host: host, port: 443, kind: .relay),
            serverID: serverID,
            pingpongPath: pingpongPath
        )
    }

    private static func decodeResponses(from data: Data) throws -> [QuickConnectResponse] {
        do {
            return try JSONDecoder().decode([QuickConnectResponse].self, from: data)
        } catch {
            throw QuickConnectResolutionError.invalidResponse
        }
    }

    private static func isTrustedDirectHost(_ host: String) -> Bool {
        let labels = host.lowercased().split(separator: ".", omittingEmptySubsequences: false)
        guard labels.count >= 4,
              labels[labels.count - 3] == "direct",
              labels[labels.count - 2] == "quickconnect",
              labels.last == "to" || labels.last == "cn" else {
            return false
        }
        return labels.allSatisfy { isValidHostLabel(String($0)) }
    }

    public static func isTrustedRelayHost(_ host: String) -> Bool {
        let labels = host.lowercased().split(separator: ".", omittingEmptySubsequences: false)
        guard labels.count == 4,
              labels[1] != "direct",
              labels[2] == "quickconnect",
              labels[3] == "to" || labels[3] == "cn" else {
            return false
        }
        return isValidHostLabel(String(labels[0])) && isValidHostLabel(String(labels[1]))
    }

    private static func isTrustedControlHost(_ host: String) -> Bool {
        (host.hasSuffix(".quickconnect.to") || host.hasSuffix(".quickconnect.cn"))
            && host.split(separator: ".").allSatisfy { isValidHostLabel(String($0)) }
    }

    private static func isValidHostLabel(_ value: String) -> Bool {
        guard (1...63).contains(value.count), value.first != "-", value.last != "-" else {
            return false
        }
        let allowed = CharacterSet(
            charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-"
        )
        return value.unicodeScalars.allSatisfy(allowed.contains)
    }
}

File: windows/src/LanStash.Infrastructure/DsmQuickConnectResolver.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using LanStash.Domain;

namespace LanStash.Infrastructure;

public enum QuickConnectEndpointKind
{
    Local,
    External,
    Relay,
}

public sealed record QuickConnectEndpoint(string Host, int Port, QuickConnectEndpointKind Kind);

/// <summary>
/// 受限使用 Synology QuickConnect 当前客户端契约。
/// get_server_info、request_tunnel 属于未公开的内部 API；登录前会限制官方域名并验证中继身份。
/// </summary>
public sealed class DsmQuickConnectResolver(HttpClient httpClient)
{
    private const int MaximumResponseBytes = 1024 * 1024;
    private readonly HttpClient _http = httpClient;
    private static readonly string[] ControlUrls =
    [
        "https://global.quickconnect.to/Serv.php",
        "https://global.quickconnect.cn/Serv.php",
    ];

    public async Task<IReadOnlyList<QuickConnectEndpoint>> ResolveAsync(
        string id,
        CancellationToken cancellationToken = default)
    {
        ValidateId(id);
        // 在线但没有直连候选时保留结果，让连接层继续中继，不能被另一区域的失败覆盖。
        return DecodeEndpoints(await QueryServerInfoAsync(id, cancellationToken).ConfigureAwait(false));
    }

    public async Task<QuickConnectEndpoint> RequestRelayAsync(
        string id,
        CancellationToken cancellationToken = default)
    {
        ValidateId(id);
        var controlUrl = await ResolveControlUrlAsync(id, cancellationToken).ConfigureAwait(false);
        DsmException? lastError = null;
        for (var attempt = 0; attempt < 3; attempt++)
        {
            try
            {
                var descriptor = DecodeRelayDescriptor(await SendAsync(
                    "request_tunnel", true, id, controlUrl, cancellationToken).ConfigureAwait(false), id);
                await VerifyRelayAsync(descriptor, cancellationToken).ConfigureAwait(false);
                return descriptor.Endpoint;
            }
            catch (DsmException error) when (
                error.Kind is not DsmErrorKind.QuickConnectRelayDisabled and
                not DsmErrorKind.QuickConnectIdentityMismatch)
            {
                lastError = error;
                if (attempt < 2)
                {
                    await Task.Delay((attempt + 1) * 1000, cancellationToken).ConfigureAwait(false);
                }
            }
        }
        throw lastError ?? RelayUnavailable();
    }

    public static IReadOnlyList<QuickConnectEndpoint> DecodeEndpoints(JsonArray responses)
    {
        var response = SuccessfulResponse(responses);
        if (!string.Equals(response["server"]?["ds_state"]?.GetValue<string>(), "CONNECTED",
                StringComparison.OrdinalIgnoreCase))
        {
            throw new DsmException(
                UserText.Key("WinSharede32c368700e2b141"),
                UserText.Key("WinSharede0abefcec6920022"),
                kind: DsmErrorKind.QuickConnectOffline);
        }
        var port = response["service"]?["port"]?.GetValue<int>() ?? 0;
        var smartDns = response["smartdns"] as JsonObject;
        if (port is < 1 or > 65535 || smartDns is null)
        {
            throw NoDirectRoute();
        }

        var endpoints = new List<QuickConnectEndpoint>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        if (smartDns["lan"] is JsonArray lan)
        {
            foreach (var value in lan)
            {
                var host = value is JsonObject item
                    ? item["host"]?.GetValue<string>()
                    : value?.GetValue<string>();
                AddTrustedDirect(endpoints, seen, host, port, QuickConnectEndpointKind.Local);
            }
        }
        AddTrustedDirect(
            endpoints,
            seen,
            smartDns["host"]?.GetValue<string>(),
            port,
            QuickConnectEndpointKind.External);
        return endpoints.Count > 0 ? endpoints : throw NoDirectRoute();
    }

    private async Task<string> ResolveControlUrlAsync(string id, CancellationToken cancellationToken)
    {
        var response = SuccessfulResponse(await QueryServerInfoAsync(id, cancellationToken).ConfigureAwait(false));
        var host = response["env"]?["control_host"]?.GetValue<string>()?.ToLowerInvariant();
        if (host is null || !IsTrustedControlHost(host)) throw InvalidResponse();
        return $"https://{host}/Serv.php";
    }

    private async Task<JsonArray> QueryServerInfoAsync(string id, CancellationToken cancellationToken)
    {
        DsmException? lastError = null;
        var pending = new Queue<string>(ControlUrls);
        var scheduled = new HashSet<string>(ControlUrls, StringComparer.OrdinalIgnoreCase);
        // 官方控制响应的 sites 是后续查询位置，不是设备不存在；限制域名、去重及总请求数。
        while (pending.TryDequeue(out var controlUrl))
        {
            cancellationToken.ThrowIfCancellationRequested();
            try
            {
                var responses = await SendAsync("get_server_info", false, id, controlUrl, cancellationToken).ConfigureAwait(false);
                if (responses.OfType<JsonObject>().Any(item => item["errno"]?.GetValue<int>() == 0)) return responses;
                foreach (var response in responses.OfType<JsonObject>())
                {
                    if (response["sites"] is not JsonArray sites) continue;
                    foreach (var site in sites)
                    {
                        if (site is not JsonValue value || !value.TryGetValue<string>(out var host) || !IsTrustedControlHost(host))
                            throw InvalidResponse();
                        var url = $"https://{host.ToLowerInvariant()}/Serv.php";
                        if (scheduled.Count < 8 && scheduled.Add(url)) pending.Enqueue(url);
                    }
                }
                _ = SuccessfulResponse(responses);
            }
            catch (DsmException error)
            {
                lastError = error;
            }
        }
        throw lastError ?? ServiceUnavailable();
    }

    private async Task<JsonArray> SendAsync(
        string command,
        bool stopWhenSuccess,
        string serverId,
        string controlUrl,
        CancellationToken cancellationToken)
    {
        var payload = new JsonArray
        {
            new JsonObject
            {
                ["version"] = 1,
                ["command"] = command,
                ["stop_when_error"] = false,
                ["stop_when_success"] = stopWhenSuccess,
                ["id"] = "mainapp_https",
                ["serverID"] = serverId,
                ["is_gofile"] = false,
                ["path"] = "",
            },
        };
        using var request = new HttpRequestMessage(HttpMethod.Post, controlUrl)
        {
            Content = new StringContent(payload.ToJsonString(), Encoding.UTF8, "application/json"),
        };
        request.Headers.Accept.ParseAdd("application/json");
        request.Headers.UserAgent.ParseAdd("LanStash-Windows/0.1");
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeout.CancelAfter(command == "request_tunnel" ? TimeSpan.FromSeconds(30) : TimeSpan.FromSeconds(15));
        try
        {
            using var response = await _http.SendAsync(
                request, HttpCompletionOption.ResponseHeadersRead, timeout.Token).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServiceUnavailable();
            }
            return await ReadJsonArrayAsync(response, timeout.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            throw ServiceUnavailable();
        }
        catch (HttpRequestException)
        {
            throw ServiceUnavailable();
        }
    }

    private async Task VerifyRelayAsync(RelayDescriptor descriptor, CancellationToken cancellationToken)
    {
        if (!descriptor.PingPongPath.StartsWith('/') ||
            descriptor.PingPongPath.Length > 2048 ||
            descriptor.PingPongPath.Contains("://", StringComparison.Ordinal) ||
            descriptor.PingPongPath.Contains('#'))
        {
            throw InvalidResponse();
        }
        var expectedId = Convert.ToHexString(
            MD5.HashData(Encoding.UTF8.GetBytes(descriptor.ServerId))).ToLowerInvariant();
        var uri = new Uri($"https://{descriptor.Endpoint.Host}{descriptor.PingPongPath}");
        for (var attempt = 0; attempt < 6; attempt++)
        {
            try
            {
                using var request = new HttpRequestMessage(HttpMethod.Get, uri);
                request.Headers.Accept.ParseAdd("application/json");
                request.Headers.UserAgent.ParseAdd("LanStash-Windows/0.1");
                using var response = await _http.SendAsync(
                    request, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
                if (!response.IsSuccessStatusCode)
                {
                    throw RelayUnavailable();
                }
                var json = await ReadJsonObjectAsync(response, cancellationToken).ConfigureAwait(false);
                if (!string.Equals(json["ezid"]?.GetValue<string>(), expectedId, StringComparison.OrdinalIgnoreCase))
                {
                    throw new DsmException(
                        UserText.Key("WinShared408b64a41c09d2a0"),
                        UserText.Key("WinShared51f93605004a3258"),
                        kind: DsmErrorKind.QuickConnectIdentityMismatch);
                }
                return;
            }
            catch (DsmException error) when (
                error.Kind != DsmErrorKind.QuickConnectIdentityMismatch && attempt < 5)
            {
                await Task.Delay(1000, cancellationToken).ConfigureAwait(false);
            }
            catch (HttpRequestException) when (attempt < 5)
            {
                await Task.Delay(1000, cancellationToken).ConfigureAwait(false);
            }
        }
        throw RelayUnavailable();
    }

    private static RelayDescriptor DecodeRelayDescriptor(JsonArray responses, string id)
    {
        if (responses.OfType<JsonObject>().Any(item => item["errno"]?.GetValue<int>() == 19))
        {
            throw new DsmException(
                UserText.Key("WinSharedd11595c8e583c9ea"),
                UserText.Key("WinShared6730f03f8272b1f9"),
                kind: DsmErrorKind.QuickConnectRelayDisabled);
        }
        var response = SuccessfulResponse(responses);
        var service = response["service"] as JsonObject;
        var server = response["server"] as JsonObject;
        var environment = response["env"] as JsonObject;
        var relayPort = service?["relay_port"]?.GetValue<int>() ?? 0;
        var serverId = server?["serverID"]?.GetValue<string>();
        var region = environment?["relay_region"]?.GetValue<string>()?.ToLowerInvariant();
        var controlHost = environment?["control_host"]?.GetValue<string>()?.ToLowerInvariant();
        if (string.IsNullOrWhiteSpace(service?["relay_ip"]?.GetValue<string>()) ||
            relayPort is < 1 or > 65535 ||
            string.IsNullOrWhiteSpace(serverId) ||
            region is null || !IsValidHostLabel(region) ||
            controlHost is null || !IsTrustedControlHost(controlHost))
        {
            throw RelayUnavailable();
        }
        var topDomain = controlHost.EndsWith(".quickconnect.cn", StringComparison.Ordinal) ? "cn" : "to";
        var host = $"{id.ToLowerInvariant()}.{region}.quickconnect.{topDomain}";
        if (!IsTrustedRelayHost(host))
        {
            throw InvalidResponse();
        }
        return new(
            new QuickConnectEndpoint(host, 443, QuickConnectEndpointKind.Relay),
            serverId,
            server?["pingpong_path"]?.GetValue<string>()
                ?? "/webman/pingpong.cgi?action=cors&quickconnect=true");
    }

    public static bool IsTrustedRelayHost(string host)
    {
        var labels = host.ToLowerInvariant().Split('.');
        return labels.Length == 4 &&
            labels[2] == "quickconnect" &&
            labels[3] is "to" or "cn" &&
            IsValidHostLabel(labels[0]) &&
            IsValidHostLabel(labels[1]);
    }

    private static JsonObject SuccessfulResponse(JsonArray responses) =>
        responses.OfType<JsonObject>().FirstOrDefault(item => item["errno"]?.GetValue<int>() == 0)
        ?? throw new DsmException(
            UserText.Key("WinShared35964d160c6c4b32"),
            UserText.Key("WinShared36d42da6d6cea95f"),
            kind: DsmErrorKind.QuickConnectNotFound);

    private static void AddTrustedDirect(
        ICollection<QuickConnectEndpoint> endpoints,
        ISet<string> seen,
        string? rawHost,
        int port,
        QuickConnectEndpointKind kind)
    {
        var host = rawHost?.ToLowerInvariant();
        if (host is not null && IsTrustedDirectHost(host) && seen.Add(host))
        {
            endpoints.Add(new(host, port, kind));
        }
    }

    private static bool IsTrustedDirectHost(string host)
    {
        var labels = host.Split('.');
        return labels.Length >= 4 &&
            labels[^3] == "direct" &&
            labels[^2] == "quickconnect" &&
            labels[^1] is "to" or "cn" &&
            labels.All(IsValidHostLabel);
    }

    private static bool IsTrustedControlHost(string host) =>
        (host.EndsWith(".quickconnect.to", StringComparison.Ordinal) ||
         host.EndsWith(".quickconnect.cn", StringComparison.Ordinal)) &&
        host.Split('.').All(IsValidHostLabel);

    private static bool IsValidHostLabel(string value) =>
        value.Length is >= 1 and <= 63 &&
        value[0] != '-' &&
        value[^1] != '-' &&
        value.All(character =>
            character is >= 'a' and <= 'z' or >= 'A' and <= 'Z' or >= '0' and <= '9' or '-');

    private static void ValidateId(string id)
    {
        if (!NasAddressParser.IsPotentialQuickConnectId(id))
        {
            throw new DsmException(
                UserText.Key("WinShared03b4e19bbedaaf48"),
                UserText.Key("WinShared89c44574bbf9988a"),
                kind: DsmErrorKind.InvalidQuickConnectId);
        }
    }

    private static async Task<JsonArray> ReadJsonArrayAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        var node = await ReadJsonAsync(response, cancellationToken).ConfigureAwait(false);
        return node as JsonArray ?? throw InvalidResponse();
    }

    private static async Task<JsonObject> ReadJsonObjectAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        var node = await ReadJsonAsync(response, cancellationToken).ConfigureAwait(false);
        return node as JsonObject ?? throw RelayUnavailable();
    }

    private static async Task<JsonNode?> ReadJsonAsync(
        HttpResponseMessage response,
        CancellationToken cancellationToken)
    {
        if (response.Content.Headers.ContentLength > MaximumResponseBytes)
        {
            throw InvalidResponse();
        }
        await using var source = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
        using var memory = new MemoryStream();
        var buffer = new byte[8192];
        while (true)
        {
            var count = await source.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
            if (count == 0)
            {
                break;
            }
            if (memory.Length + count > MaximumResponseBytes)
            {
                throw InvalidResponse();
            }
            await memory.WriteAsync(buffer.AsMemory(0, count), cancellationToken).ConfigureAwait(false);
        }
        memory.Position = 0;
        return await JsonNode.ParseAsync(memory, cancellationToken: cancellationToken).ConfigureAwait(false);
    }

    private static DsmException NoDirectRoute() => new(
        UserText.Key("WinShared2185bc1f61ec5011"),
        UserText.Key("WinSharedc157df0fa964b080"),
        kind: DsmErrorKind.QuickConnectDirectUnavailable);
    private static DsmException ServiceUnavailable() => new(
        UserText.Key("WinShared5634da8d0d5a2089"),
        UserText.Key("WinSharedefc81ced18eb3bb0"),
        kind: DsmErrorKind.QuickConnectServiceUnavailable);
    private static DsmException InvalidResponse() => new(
        UserText.Key("WinSharedda35e58bcad31766"),
        UserText.Key("WinSharedefc81ced18eb3bb0"),
        kind: DsmErrorKind.QuickConnectInvalidResponse);
    private static DsmException RelayUnavailable() => new(
        UserText.Key("WinShared165059223c7f83c1"),
        UserText.Key("WinSharedefc81ced18eb3bb0"),
        kind: DsmErrorKind.QuickConnectRelayUnavailable);

    private sealed record RelayDescriptor(
        QuickConnectEndpoint Endpoint,
        string ServerId,
        string PingPongPath);
}
